src/main/java/org/prebid/server/bidder/huaweiads/HuaweiAdSlotBuilder.java

```diff
@@ -141,13 +147,80 @@ private List<Asset> parseNativeRequestAssets(Native xNative) {
     private static List<String> makeDetailedCreativeTypeList(long numVideo, long numImage) {
         if (numVideo >= 1) {
             return List.of("903");
-        } else if (numImage > 1) {
-            return List.of("904");
-        } else if (numImage == 1) {
-            return List.of("901");
+        } else if (numImage >= 1) {
+            return List.of("901", "904", "905");
         } else {
             return List.of("913", "914");
         }
     }
 
+    private List<Format> makeFormatListForNative(List<Asset> assets, long numImage) {
+        final Set<Format> formats = new HashSet<>();
+        formats.addAll(makeFormatListForVideo(assets));
+        formats.addAll(makeFormatListForImage(assets, numImage));
+        return formats.stream().toList();
+    }
+
+    private List<Format> makeFormatListForVideo(List<Asset> assets) {
+        return assets.stream()
+                .map(Asset::getVideo)
+                .filter(Objects::nonNull)
+                .filter(video -> HuaweiUtils.isFormatDefined(video.getW(), video.getH()))
+                .map(video -> Format.of(video.getW(), video.getH()))
+                .distinct()
+                .toList();
+    }
+
+    private Set<Format> makeFormatListForImage(List<Asset> assets, long numImage) {
+        final Set<Format> formats = new HashSet<>();
+        assets.stream()
+                .map(Asset::getImg)
+                .filter(Objects::nonNull)
+                .filter(image -> HuaweiUtils.isFormatDefined(image.getW(), image.getW()))
```

Comment:
typo: the `getW()` is called twice, as I understand it should be `HuaweiUtils.isFormatDefined(image.getWmin(), image.getHmin()`, than you can simplify the following `foreach` logic

Reply:
> typo: the getW() is called twice
Oh, thx!

> than you can simplify the following foreach logic
Hmm... Filter should be removed entirely, actually.